### ISSUE TITLE
RangeError: Trying to access beyond buffer length

### ISSUE THREAD

**Issue by yodasan** [bitcoinjs/bitcoinjs-lib#344]
I am having an issue where we get a stack trace thrown when we load up multi-sig transactions from a large amount of inputs. I altered the "works for the P2SH multisig case" test case to demonstrate. 

Stack trace:
```
RangeError: Trying to access beyond buffer length
      at checkOffset (buffer.js:582:11)
      at Buffer.readUInt32LE (buffer.js:642:5)
      at Object.readPushDataInt (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/bufferutils.js:42:21)
      at Function.Script.fromBuffer (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/script.js:40:27)
      at isScriptHashInput (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/scripts.js:73:29)
      at classifyInput (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/scripts.js:142:14)
      at isScriptHashInput (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/scripts.js:75:10)
      at Object.classifyInput (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/scripts.js:142:14)
      at /Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/transaction_builder.js:47:30
      at Array.forEach (native)
      at Function.TransactionBuilder.fromTransaction (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/src/transaction_builder.js:37:19)
      at Context.<anonymous> (/Users/yodasan/bitcoin-dev/bitcoinjs-lib/test/transaction_builder.js:325:39)
      at callFn (/usr/local/lib/node_modules/mocha/lib/runnable.js:251:21)
      at Test.Runnable.run (/usr/local/lib/node_modules/mocha/lib/runnable.js:244:7)
      at Runner.runTest (/usr/local/lib/node_modules/mocha/lib/runner.js:374:10)
      at /usr/local/lib/node_modules/mocha/lib/runner.js:452:12
      at next (/usr/local/lib/node_modules/mocha/lib/runner.js:299:14)
      at /usr/local/lib/node_modules/mocha/lib/runner.js:309:7
      at next (/usr/local/lib/node_modules/mocha/lib/runner.js:248:23)
      at Object._onImmediate (/usr/local/lib/node_modules/mocha/lib/runner.js:276:5)
      at processImmediate [as _immediateCallback] (timers.js:330:15)
```

```javascript
it('reading large multi-sig transaction from hex', function() {
        var privKeys = [
            "91avARGdfge8E4tZfYLoxeJ5sGBdNJQH4kvjJoQFacbgwmaKkrx",
            "91avARGdfge8E4tZfYLoxeJ5sGBdNJQH4kvjJoQFacbgww7vXtT"
        ].map(ECKey.fromWIF)
        var redeemScript = Script.fromASM("OP_2 0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8 04c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee51ae168fea63dc339a3c58419466ceaeef7f632653266d0e1236431a950cfe52a OP_2 OP_CHECKMULTISIG")

        var numberOfInputs = 300;

        for(var i=0; i < numberOfInputs; i++) {
            txb.addInput("4971f016798a167331bcbc67248313fbc444c6e92e4416efd06964425588f5cf", i);
        }

        txb.addOutput("1BgGZ9tcN4rm9KBzDn7KprQz87SZ26SAMH", 10000)

        for(i=0; i < numberOfInputs; i++) {
            txb.sign(i, privKeys[0], redeemScript)
        }


        var tx = txb.buildIncomplete()
        var hexOutputOne = tx.toHex();

        var txb2 = TransactionBuilder.fromTransaction(tx)
    })
```

**Comment by dcousens** [comment]
Can confirm this.  Will investigate, cheers for reporting this @yodasan

**Comment by dcousens** [comment]
Still investigating this... will have the test fixture added to the library soon.

**Comment by dcousens** [comment]
@yodasan this was fixed by https://github.com/bitcoinjs/bitcoinjs-lib/pull/354, but I haven't isolated the problem yet... still investigating.

**Comment by dcousens** [comment]
Ok, figured it out!  What a nuisance.

Because the signature is arbitrary data until treated as a signature, `isScriptHashInput` during classification is attempting to check if it is a valid script by doing [`Script.fromBuffer`](https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/scripts.js#L73), since it is a signature, it is naturally exploding because `304...` is not an encoded variable length integer.

It was resolved in #354 because the [order of evaluation was changed](https://github.com/bitcoinjs/bitcoinjs-lib/pull/354/files#diff-dce2d65bc8a0d0097775001513aae5c5L142),  such that it is classified as multisignature first.

This just happened to be a unique set of circumstance that brought it about.

**Comment by dcousens** [comment]
To fix this, I need to fix these assumptions in the `is*` functions, and add respective invalid tests so it doesn't explode implementations.
